cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/HttpRequestHandler.java

```diff
@@ -66,6 +66,7 @@
   private final Map<WrappedDiscoverable, MessageSender> discoveryLookup;
   private final List<ProxyRule> proxyRules;
 
+  private boolean exceptionHandled = false;
```

Comment:
This needs to be volatile, if `exceptionCaught` can be called from multiple threads?